src/Service/EolValidator.php

```diff
@@ -0,0 +1,163 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\MagentoCloud\Service;
+
+use Carbon\Carbon;
+use Composer\Semver\Semver;
+use Magento\MagentoCloud\Config\ValidatorInterface;
+use Magento\MagentoCloud\Filesystem\Driver\File;
+use Magento\MagentoCloud\Filesystem\FileList;
+use Magento\MagentoCloud\Filesystem\FileSystemException;
+use Symfony\Component\Yaml\Yaml;
+
+/**
+ * Service EOL validator.
+ *
+ * Class EolValidator
+ */
+class EolValidator
+{
+    /**
+     * Set the notification period.
+     */
+    private const NOTIFICATION_PERIOD = 3;
+
+    /**
+     * @var FileList
+     */
+    private $fileList;
+
+    /**
+     * @var File
+     */
+    private $file;
+
+    /**
+     * @var ServiceFactory
+     */
+    private $serviceFactory;
+
+    /**
+     * @var array
+     */
+    private $eolConfigs;
+
+    /**
+     * @var array
+     */
+    private $services = [
+        ServiceInterface::NAME_PHP,
+        ServiceInterface::NAME_ELASTICSEARCH,
+        ServiceInterface::NAME_RABBITMQ,
+        ServiceInterface::NAME_REDIS,
+        ServiceInterface::NAME_DB
+    ];
+
+    /**
+     * @param FileList $fileList
+     * @param ServiceFactory $serviceFactory
+     */
+    public function __construct(
+        FileList $fileList,
+        File $file,
+        ServiceFactory $serviceFactory
+    ) {
+        $this->fileList = $fileList;
+        $this->file = $file;
+        $this->serviceFactory = $serviceFactory;
+    }
+
+    /**
+     * Validate the EOL of a given service and version by error level.
+     *
+     * @return array
+     * @throws FileSystemException
+     * @throws ServiceMismatchException
+     */
+    public function validateServiceEol(): array
+    {
+        $errors = [];
+
+        foreach ($this->services as $serviceName) {
+            $service = $this->serviceFactory->create($serviceName);
+            $serviceVersion = $service->getVersion();
+
+            if ($validationResult = $this->validateService($serviceName, $serviceVersion)) {
+                $errorLevel = array_key_first($validationResult);
+                $errors[$errorLevel][] = $validationResult[$errorLevel];
+            }
+        }
+        return $errors;
+    }
+
+    /**
+     * Validates a given service and version.
+     *
+     * @param string $serviceName
+     * @param string $serviceVersion
+     * @return array
+     * @throws FileSystemException
+     */
+    public function validateService(string $serviceName, string $serviceVersion) : array
+    {
+        $serviceConfigs = $this->getServiceConfigs($serviceName);
+
+        $versionConfigs = array_filter($serviceConfigs, function ($v) use ($serviceVersion) {
+            return Semver::satisfies($serviceVersion, sprintf('%s.x', $v['version']));
+        });
+
+        if (!$versionConfigs || $versionConfigs[array_key_first($versionConfigs)]['eol'] === null) {
```

Comment:
If I understand correctly: according to this condition `$versionConfigs[array_key_first($versionConfigs)]['eol'] === null)` for Redis service it always will be `true` as eol isn't defined. And always will be returned warning level error .

Is it correct behavior?

Reply:
Correct. That will return a warning with a message such as 'Unknown or invalid EOL defined for redis <version>'.

Reply:
But why do we need to display such error if eol isn't defined? For users it can be confused. 
Maybe we shouldn't return an error in such cases?

Reply:
Updated.